src/Nethermind/Nethermind.Core/Collections/CollectionExtensions.cs

```diff
@@ -22,5 +29,61 @@ public static void AddRange<T>(this ICollection<T> list, params T[] items)
                 list.Add(items[index]);
             }
         }
+
+        public static void NoResizeClear<TKey, TValue>(this ConcurrentDictionary<TKey, TValue> dictionary)
+                where TKey : notnull
+        {
+            if (dictionary is null || dictionary.IsEmpty)
+            {
+                return;
+            }
+
+            using var handle = dictionary.AcquireLock();
+
+            ClearCache<TKey, TValue>.Clear(dictionary);
+        }
+
+        private static class ClearCache<TKey, TValue> where TKey : notnull
+        {
+            public static readonly Action<ConcurrentDictionary<TKey, TValue>> Clear = CreateNoResizeClearExpression();
+
+            private static Action<ConcurrentDictionary<TKey, TValue>> CreateNoResizeClearExpression()
+            {
+                // Parameters
+                var dictionaryParam = Expression.Parameter(typeof(ConcurrentDictionary<TKey, TValue>), "dictionary");
+
+                // Access _tables field
+                var tablesField = typeof(ConcurrentDictionary<TKey, TValue>).GetField("_tables", BindingFlags.NonPublic | BindingFlags.Instance);
+                var tablesAccess = Expression.Field(dictionaryParam, tablesField!);
+
+                // Access _buckets and _countPerLock fields
+                var tablesType = tablesField!.FieldType;
+                var bucketsField = tablesType.GetField("_buckets", BindingFlags.NonPublic | BindingFlags.Instance);
+                var countPerLockField = tablesType.GetField("_countPerLock", BindingFlags.NonPublic | BindingFlags.Instance);
+
+                var bucketsAccess = Expression.Field(tablesAccess, bucketsField!);
+                var countPerLockAccess = Expression.Field(tablesAccess, countPerLockField!);
+
+                // Clear arrays using Array.Clear
+                var clearMethod = typeof(Array).GetMethod("Clear", new[] { typeof(Array), typeof(int), typeof(int) });
+
+                var clearBuckets = Expression.Call(clearMethod!,
+                    bucketsAccess,
+                    Expression.Constant(0),
+                    Expression.ArrayLength(bucketsAccess));
+
+                var clearCountPerLock = Expression.Call(clearMethod!,
+                    countPerLockAccess,
+                    Expression.Constant(0),
+                    Expression.ArrayLength(countPerLockAccess));
+
+                // Block to execute both clears
+                var block = Expression.Block(clearBuckets, clearCountPerLock);
+
+                // Compile the expression into a lambda
+                return Expression.Lambda<Action<ConcurrentDictionary<TKey, TValue>>>(block, dictionaryParam).Compile();
+            }
```

Comment:
Nice hack FTW!

Can we add a test for it? For example if we upgrade to .Net X and internals change the test should fail.

Reply:
Added test